### PR TITLE
HIVE-24484: Upgrade Hadoop to 3.3.1

### PR DESCRIPTION
**Pull request by belugabehr**

### CODE REVIEW

**Comment by wangyum** [comment]
Any update?

**Comment by belugabehr** [comment]
@wangyum Stuck.

There are two big issues here:

1. Hive integration tests fire up Druid, Kafka, HDFS, LLAP, etc. all in the same JVM and their 3rd party dependencies are all over the place. Using a higher version of a dependency breaks one product, but using a lower version breaks the other.  To make this work well, there probably needs to be a way to launch each service in their own JVM class loader.  In lieu of that, I've been trying to move the ball closer to the goal post and getting dependencies closer together.

https://github.com/apache/druid/pull/10683
HIVE-24542


2. In HDFS 3.3.0, Hadoop team introduced `ProtobufRpcEngine2` in addition to `ProtobufRpcEngine` (sigh).  Some of the Hive LLAP stuff is using this Hadoop Protobuf RPC engine (`ProtobufRpcEngine`).  There's some `static` logic in the protocol engines that prohibits loading both RPC engines into the same JVM at the same time, I'm not sure why.  HDFS was migrated to `ProtobufRpcEngine2`.  So, again, in the integration tests, when the HDFS mini cluster is loaded, version 2 of the RPC engine is loaded into the JVM.  When LLAP is later loaded, it fails to start because version 1 cannot be registered at the same time.

**Comment by vyaslav** [comment]
> @wangyum Stuck.
> 
> There are two big issues here:
> 
> 1. Hive integration tests fire up Druid, Kafka, HDFS, LLAP, etc. all in the same JVM and their 3rd party dependencies are all over the place. Using a higher version of a dependency breaks one product, but using a lower version breaks the other.  To make this work well, there probably needs to be a way to launch each service in their own JVM class loader.  In lieu of that, I've been trying to move the ball closer to the goal post and getting dependencies closer together.
> 
> [apache/druid#10683](https://github.com/apache/druid/pull/10683)
> [HIVE-24542](https://issues.apache.org/jira/browse/HIVE-24542)
> 
> 1. In HDFS 3.3.0, Hadoop team introduced `ProtobufRpcEngine2` in addition to `ProtobufRpcEngine` (sigh).  Some of the Hive LLAP stuff is using this Hadoop Protobuf RPC engine (`ProtobufRpcEngine`).  There's some `static` logic in the protocol engines that prohibits loading both RPC engines into the same JVM at the same time, I'm not sure why.  HDFS was migrated to `ProtobufRpcEngine2`.  So, again, in the integration tests, when the HDFS mini cluster is loaded, version 2 of the RPC engine is loaded into the JVM.  When LLAP is later loaded, it fails to start because version 1 cannot be registered at the same time.

Regarding 1st, I faced the same issues in my PR for upgrade to 3.1.3 - https://github.com/apache/hive/pull/1638
But, regarding 2nd I'am curious if it would be hard to replace `ProtobufRpcEngine` with `ProtobufRpcEngine2` in Hive itself. As I understand they have upgraded from PB2 to PB3

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java [hunk 1]
```diff
@@ -119,16 +120,15 @@ public void targetAndSourceHaveDifferentEncryptionZoneKeys() throws Throwable {
...
-        .run("select value from encrypted_table")
-        .verifyFailure(new String[] { "value1", "value2" });
+    try {
+      replica.run("repl load " + primaryDbName + " into " + replicatedDbName
+          + " with('hive.repl.add.raw.reserved.namespace'='true', " + "'hive.repl.replica.external.table.base.dir'='"
+          + replica.externalTableWarehouseRoot + "', "
+          + "'distcp.options.pugpbx'='', 'distcp.options.skipcrccheck'='')");
+      Assert.fail("Test should have thrown an exception because cross-encryption-zone is not allowed for RAW");
+    } catch (IOException ioe) {
+      // ignore
```
Comment (pgaref) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java:130]: Hey @belugabehr  just read the detailed comment on the JIRA about this but I believe we should add some explanation here as well for clarity

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapInputFormat.java [hunk 2]
```diff
@@ -137,6 +137,8 @@
       // This starts the reader in the background.
       rr.start();
       return result;
+    } catch (IOException ioe) {
```
Comment (pgaref) [inline llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapInputFormat.java:140]: is this needed? Exception should catch everything right?

File: ql/src/test/results/clientpositive/llap/check_constraint.q.out [hunk 3]
```diff
@@ -2415,14 +2415,14 @@ STAGE PLANS:
                 outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5
                 Statistics: Num rows: 1 Data size: 409 Basic stats: COMPLETE Column stats: NONE
                 Select Operator
-                  expressions: _col1 (type: string), _col0 (type: int), _col4 (type: string), _col5 (type: struct<writeid:bigint,bucketid:int,rowid:bigint>), _col2 (type: string), _col3 (type: int)
+                  expressions: _col0 (type: int), _col4 (type: string), _col3 (type: int), _col1 (type: string), _col5 (type: struct<writeid:bigint,bucketid:int,rowid:bigint>), _col2 (type: string)
```
Comment (pgaref) [inline ql/src/test/results/clientpositive/llap/check_constraint.q.out:2418]: Are these q.out changes expected? What is the root cause?

File: llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapInputFormat.java (hunk 2 above)
Comment (belugabehr) [reply llap-server/src/java/org/apache/hadoop/hive/llap/io/api/impl/LlapInputFormat.java:140]: Hey @pgaref,

Ya, this is required.  Based on the `InvalidInputException` (which is a subclass of `IOException`) changes in HDFS, this code is require to pass the `InvalidInputException` up to the caller directly, otherwise, in the `Exception` block, it gets wrapped in yet another `IOException` and that caller is no longer able to detect the `InvalidInputException`.

I hope that makes sense.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java [hunk 4]
```diff
@@ -55,14 +56,15 @@ static void internalBeforeClassSetup(Map<String, String> overrides, Class clazz)
       throws Exception {
     conf = new HiveConf(clazz);
     conf.set("dfs.client.use.datanode.hostname", "true");
-    conf.set("hadoop.proxyuser." + Utils.getUGI().getShortUserName() + ".hosts", "*");
```
Comment (abstractdog) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java:58]: this is for impersonation according to https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/Superusers.html, don't we want to test this scenario anymore?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 5]
```diff
@@ -4117,28 +4118,33 @@ public void testAuthForNotificationAPIs() throws Exception {
...
+    // Remove the proxy privilege by reseting proxy configuration to default value.
+    // The auth should fail (in reality the proxy setting should not be changed on the fly)
+    // Pretty hacky: Affects both instances of HMS
+    ProxyUsers.refreshSuperUserGroupsConfiguration();
+
     try {
       rsp = metaStoreClient.getNextNotification(firstEventId, 0, null);
+      Assert.fail("Get Next Nofitication should have failed due to no proxy auth");
     } catch (TException e) {
-      ex = e;
```
Comment (abstractdog) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:4130]: I have no idea how can we hit this catch, but having it empty is always a red sign, have you checked if at least a log.debug is useful here?

File: ql/src/java/org/apache/hadoop/hive/ql/io/RecordReaderWrapper.java [hunk 6]
```diff
@@ -69,7 +70,14 @@ static RecordReader create(InputFormat inputFormat, HiveInputFormat.HiveInputSpl
...
     int headerCount = Utilities.getHeaderCount(tableDesc);
     int footerCount = Utilities.getFooterCount(tableDesc, jobConf);
-    RecordReader innerReader = inputFormat.getRecordReader(split.getInputSplit(), jobConf, reporter);
+
+    RecordReader innerReader = null;
+    try {
+     innerReader = inputFormat.getRecordReader(split.getInputSplit(), jobConf, reporter);
+    } catch (InterruptedIOException iioe) {
+      // If reading from the underlying record reader is interrupted, return a no-op record reader
+      return new ZeroRowsInputFormat().getRecordReader(split.getInputSplit(), jobConf, reporter);
```
Comment (abstractdog) [inline ql/src/java/org/apache/hadoop/hive/ql/io/RecordReaderWrapper.java:79]: why is it better to return with no-op record reader instead of letting this codepath fail and handle the exception somewhere else? doesn't this mask issues?

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java [hunk 7]
```diff
@@ -483,7 +484,8 @@ HiveAuthorizer createHiveMetaStoreAuthorizer() throws Exception {
   boolean isSuperUser(String userName) {
     Configuration conf      = getConf();
     String        ipAddress = HMSHandler.getIPAddress();
-    return (MetaStoreServerUtils.checkUserHasHostProxyPrivileges(userName, conf, ipAddress));
+    ProxyUsers.refreshSuperUserGroupsConfiguration(conf);
+    return (MetaStoreServerUtils.checkUserHasHostProxyPrivileges(userName, ipAddress));
```
Comment (abstractdog) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java:488]: nit: extra bracket is not needed I guess

File: spark-client/pom.xml [hunk 8]
```diff
@@ -159,45 +159,10 @@
 
   <build>
     <plugins>
-      <plugin>
```
Comment (abstractdog) [inline spark-client/pom.xml:162]: happy to see that we can get rid of these maven magics!

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 9]
```diff
@@ -367,6 +368,7 @@ public static void startMetaStore(int port, HadoopThriftAuthBridge bridge,
     boolean tcpKeepAlive = MetastoreConf.getBoolVar(conf, ConfVars.TCP_KEEP_ALIVE);
     boolean useCompactProtocol = MetastoreConf.getBoolVar(conf, ConfVars.USE_THRIFT_COMPACT_PROTOCOL);
     boolean useSSL = MetastoreConf.getBoolVar(conf, ConfVars.USE_SSL);
+    ProxyUsers.refreshSuperUserGroupsConfiguration(conf);
```
Comment (abstractdog) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:371]: is done somewhere else implicitly before hadoop 3.3?

File: standalone-metastore/pom.xml [hunk 10]
```diff
@@ -79,8 +79,8 @@
     <dropwizard-metrics-hadoop-metrics2-reporter.version>0.1.2
     </dropwizard-metrics-hadoop-metrics2-reporter.version>
     <dropwizard.version>3.1.0</dropwizard.version>
-    <guava.version>19.0</guava.version>
-    <hadoop.version>3.1.0</hadoop.version>
+    <guava.version>27.0-jre</guava.version>
+    <hadoop.version>3.2.1</hadoop.version>
```
Comment (abstractdog) [inline standalone-metastore/pom.xml:83]: I think we're targeting 3.3.1 here too, right?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java (hunk 4 above)
Comment (belugabehr) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java:58]: Hey @abstractdog, thanks for the review.

Take a look at my notes here:

https://issues.apache.org/jira/browse/HIVE-24484?focusedCommentId=17369708&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17369708

tldr; These unit tests are launching two HMS within the same JVM (same class-loader) and therefore they are able to modify each other's state where it stored in static variables.  This testing cannot be done any more.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java (hunk 5 above)
Comment (belugabehr) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:4130]: The idea here is that it SHOULD throw an Exception.  If it does not throw an Exception from `getNextNofitication` then it will hit the `Assert.fail`.  I can add a comment to clarify that this is the expected behavior.

File: ql/src/java/org/apache/hadoop/hive/ql/io/RecordReaderWrapper.java (hunk 6 above)
Comment (belugabehr) [reply ql/src/java/org/apache/hadoop/hive/ql/io/RecordReaderWrapper.java:79]: Hey.

So, in my experimentation, this is the least-bad option.  I did this to preserve the previous behavior.  The Hive code is not setup to handle this error condition.  As thing currently stand in `master`, if the calling Thread was interrupted, the thread would finish fetching the rows regardless and then just later ignore them (throw them away).  The calling code does not handle 'null' return value and it does not handle this Exception.  As currently implemented in Hive `master`, if it gets an exception it simply exits execution with an Error message, without implementing a lot more code, there is no way to ignore/skip this one specific error type.  So, the cleanest thing to do is to return `ZeroRows` since it's going to be thrown away later anyway.

File: standalone-metastore/pom.xml (hunk 10 above)
Comment (belugabehr) [reply standalone-metastore/pom.xml:83]: Wow, great catch.  Nooooooooooo!  Ugh.

I hope it doesn't break anything.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java (hunk 9 above)
Comment (belugabehr) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:371]: So ya, this was done as a separate thing buried in the Hive code.  This makes it much more explicit and less hidden.

Before Hadoop 3.3, it could easily be detected if a call to `refreshSuperUserGroupsConfiguration` had already been performed because there was a corresponding getter that would return a `null` value if it had not.  Well, in 3.3 that went away and instead of returning null, you get some sort of default value.  So now one can't lazily refresh these configurations, if they haven't already been, it's better to just refresh them explicitly here and be done with it.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/WarehouseInstance.java [hunk 11]
```diff
@@ -394,16 +393,13 @@ WarehouseInstance verifyResults(String[] data) throws IOException {
...
-      }
-    }
-    assertFalse(dataMatched);
+    final List<String> expectedResults =
+        Arrays.asList(data).stream().map(r -> r.toLowerCase()).collect(Collectors.toList());
+    final List<String> actualResults = getOutput().stream().map(r -> r.toLowerCase()).collect(Collectors.toList());
+
+    assertTrue("Data " + expectedResults + " should not be present in " + actualResults,
+        Collections.disjoint(expectedResults, actualResults));
+
```
Comment (kgyrtkirk) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/WarehouseInstance.java:402]: old and new code seem to be doing  different things
old is:
```
e[0] != r[0] || ... || e[n] != r[n]
```
new block is a condition which is checking a set operation between the elements...

why is this change necessary - how this is connected to a hadoop upgrade?

File: ql/src/test/results/clientpositive/llap/check_constraint.q.out (hunk 3 above)
Comment (kgyrtkirk) [reply ql/src/test/results/clientpositive/llap/check_constraint.q.out:2418]: the join operand order seem to have changed

**Comment by ayushtkn** [comment]
Seems the tests are failing with
``
java.lang.NoSuchMethodError: org.jline.reader.impl.completer.StringsCompleter.<init>([Lorg/jline/reader/Candidate;)V
``
Should be fixable in the Hive Code itself?
If there is something required in the Hadoop Code, we can get that in now, 3.3.2 release is being planned out

**Comment by belugabehr** [comment]
@ayushtkn I'm not 100% sure what's going on there.  I am working on upgrading jline as a separate task HIVE-25495 (#2617) and I'm hitting a similar issue there event though I thought I synchronized the versions between Hive and Hadoop, so I need to look into it more

**Comment by belugabehr** [comment]
OK, so the issue now is that Hadoop 3.3+ is using `org.jline v3` and one of the Hive dependencies of `sqline` has a dependency on `jline v2` which is causing a clash.

**Comment by kgyrtkirk** [comment]
this PR is not making much progress - I think this in its current form will not work; or will not land soon:\
I think it would make sesnse to consider:
* split this thing up into some pieces which we could get in...
* or even... upgrade to hadoop 3.1.MAX or 3.2.ANYTHING to grab some of the changes we have to cover in upgrading to 3.3.1
instead of waiting this thing to get in with JDK11 support and everything? - what do you guys think?

**Comment by abstractdog** [comment]
what are the unresolved blockers of 3.3.1 upgrade at the moment?

**Comment by kgyrtkirk** [comment]
good question.... I don't know; but it seems like this PR have stalled!

this patch have:
* added a lib named `jansi` => this seems like an independent step
* removed the explicit guava usage from druid => could we upgrade guava separately? probably not...
* jline / jetty related things
* lots of other things

**Comment by ayushtkn** [comment]
> or even... upgrade to hadoop 3.1.MAX or 3.2.ANYTHING to grab some of the changes we have to cover in upgrading to 3.3.1

Hadoop 3.1 line is EOL and 3.1.MAX & 3.2.Anything doesn't have Guava Shaded. So, guava version mismatch leads to even more issues,
From 3.3 only Guava & Protobuf are shaded in hadoop.

> so the issue now is that Hadoop 3.3+ is using org.jline v3 and one of the Hive dependencies of sqline has a dependency on jline v2 which is causing a clash

If I decode @belugabehr's comment. The problem is only with Jline. If Jline v3 & v2 are compatible, and I see Jline is used by  only Yarn-Client, So can we just not exclude the jline dependency while adding yarn-client?

**Comment by kgyrtkirk** [comment]
@ayushtkn: I've done some digging into the jline3 issue in #2617 ([here](https://github.com/apache/hive/pull/2617#issuecomment-978029623))  and I'm not sure if it was deliberate move to declare jline3 as a dependency of the `hadoop-yarn-client` artifact
